### PR TITLE
Created AppManager class

### PR DESCRIPTION
**Pull request by albertoursino**
This class is useful for managing all those previously merged.

It allows you to **send RingCommands and manage those received**.
Through this class you can also, thanks to the "checkPassword" method, check if the password contained in the RingCommand corresponds to the one saved in memory.

### CODE REVIEW

File: app/src/main/java/com/eis/phoneringer/AppManager.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * If the password of the message received is valid then play ringtone for fixed amount of time
+     *
+     * @param context     of the application
+     * @param ringCommand a ring command not null
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    public void onRingCommandReceived(Context context, RingCommand ringCommand, final Ringtone ringtone) {
```
Comment (CremaLuca) [inline app/src/main/java/com/eis/phoneringer/AppManager.java]: If you really **really** want the `ringCommand` to be not null you could add the @NonNull annotation. It's up to you, not necessary (you could also add it on context)

File: app/src/main/java/com/eis/phoneringer/AppManager.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        if (instance == null)
+            instance = new AppManager();
+        return instance;
+    }
+
+    /**
+     * If the password of the message received is valid then play ringtone for fixed amount of time
+     *
+     * @param context     of the application
+     * @param ringCommand a ring command not null
```
Comment (CremaLuca) [inline app/src/main/java/com/eis/phoneringer/AppManager.java]: Missing `ringtone` @param description

File: app/src/main/java/com/eis/phoneringer/AppManager.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    ringtoneHandler.stopRingtone(ringtone);
+                }
+            }, TIMEOUT_TIME);
+        } else {
+            Toast.makeText(context, WRONG_PASSWORD_TOAST, Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    /**
+     * Method used to send a RingCommand via SMS using the library class {@link SMSHandler}
```
Comment (CremaLuca) [inline app/src/main/java/com/eis/phoneringer/AppManager.java]: Love the {@link}

File: app/src/main/java/com/eis/phoneringer/AppManager.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * Checks if the RingCommand password and the one saved in memory corresponds
+     *
+     * @param context     a valid context
+     * @param ringCommand a valid RingCommand object
+     * @return a boolean: true = passwords are corresponding, false = passwords are NOT corresponding
+     */
+    public boolean checkPassword(Context context, RingCommand ringCommand) {
```
Comment (CremaLuca) [inline app/src/main/java/com/eis/phoneringer/AppManager.java]: Is this method **_really_** `public`? Can't this become `private`?

File: app/src/main/java/com/eis/phoneringer/AppManager.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        }
+    }
+
+    /**
+     * Method used to send a RingCommand via SMS using the library class {@link SMSHandler}
+     *
+     * @param context     of the application
+     * @param ringCommand to send
+     * @throws InvalidTelephoneNumberException
+     * @throws InvalidSMSMessageException
```
Comment (CremaLuca) [inline app/src/main/java/com/eis/phoneringer/AppManager.java]: Are these two exceptions actually thrown? `parseCommand` doesn't throw anything, even though it probably should because we don't check `SMSPeer` validity in RingCommand

Comment (CremaLuca) [reply app/src/main/java/com/eis/phoneringer/AppManager.java]: You should probably go and edit RingCommandHandler.parseCommand and add `throws InvalidTelephoneNumberException, InvalidSMSMessageException`

**Review by CremaLuca — CHANGES_REQUESTED** [review]
A few changes, mostly for code reliability

File: app/src/main/java/com/eis/phoneringer/AppManager.java (hunk 4 above)
Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/AppManager.java]: You may be right.
I am not very convinced because it is a particular method that could also be used in other classes, but for now I put it as private as you suggest because it is actually used only within AppManager.

File: app/src/main/java/com/eis/phoneringer/AppManager.java (hunk 5 above)
Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/AppManager.java]: These exceptions can only be launched in this method because the SMSHandler is invoked only within it.

File: app/src/main/java/com/eis/phoneringer/AppManager.java (hunk 3 above)
Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/AppManager.java]: Who knows from whom I learned it.........

File: app/src/main/java/com/eis/phoneringer/AppManager.java (hunk 5 above)
Comment (CremaLuca) [reply app/src/main/java/com/eis/phoneringer/AppManager.java]: `SMSHandler` doesn't throw any exception. Those who throw those two exceptions are `SMSPeer` and `SMSMessage`, which are created inside `RingCommandHandler.parseCommand`, so that method should throw them.

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/AppManager.java]: Ops, you're right.
Is the _ringCommandhandler_ class, when creating the SMSMessage, which could launch these exceptions.

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/AppManager.java]: Well we have to update the RingCommandHandler class.

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        }
+    }
+
+    /**
+     * Method used to send a RingCommand via SMS using the library class {@link SMSHandler}
+     *
+     * @param context     of the application
+     * @param ringCommand to send
+     */
+    public void sendCommand(Context context, @NonNull RingCommand ringCommand) throws InvalidSMSMessageException, InvalidTelephoneNumberException {
```
Comment (CremaLuca) [inline app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Nah man, the two exceptions are still being thrown because this method calls `RingCommandHandler.parseCommand()` which throws them.
Also, you removed the specifications but not the "throws" after the method signature. Bring the specifications back and give them a description of when they are thrown

**Review by CremaLuca — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java (hunk 6 above)
Comment (AlbertoUrsino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: I know, corrected :D

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Ok I removed the "throws" but wait, why should I add the exceptions back and specify that they are launched by the method of **another class**?

Comment (CremaLuca) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Yes, you should specify that those exceptions are thrown because you make a call to `RingCommandHandler.parseCommand()` and that may throw exceptions

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Well ok, I put the exceptions back.
Do we leave the RingCommandHandler class as it is?

Comment (CremaLuca) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: For now yes, you can't edit it from this branch

**Review by CremaLuca — APPROVED** [review]

**Review by Cogno-Marco — APPROVED** [review]
Approving changes, quick little class not doing much 👍

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import android.os.Handler;
+import android.os.Looper;
+import android.widget.Toast;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
+
+import com.eis.smslibrary.SMSHandler;
+import com.eis.smslibrary.exceptions.InvalidSMSMessageException;
+import com.eis.smslibrary.exceptions.InvalidTelephoneNumberException;
```
Comment (giovannivelludo) [inline app/src/main/java/com/eis/phoneringer/structure/AppManager.java:13]: You're importing classes from package `com.eis.smslibrary` (which is located in another module), but you didn't include its module as a dependency in `app`'s `build.gradle`.

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * If the password of the RingCommand received is valid then play ringtone for fixed amount of time
+     *
+     * @param context     of the application
+     * @param ringCommand received
+     * @param ringtone    to be played
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    public void onRingCommandReceived(Context context, @NonNull RingCommand ringCommand, final Ringtone ringtone) {
+        final RingtoneHandler ringtoneHandler = RingtoneHandler.getInstance();
```
Comment (giovannivelludo) [inline app/src/main/java/com/eis/phoneringer/structure/AppManager.java:55]: Class RingtoneHandler does not exist in the [branch](https://github.com/AlbertoUrsino/killer-app/tree/03d65817ed7d2879e3c0e9595db7794fd9f0f3e0) you're requesting a pull from, how could you check if your code even compiles?

**Review by giovannivelludo — CHANGES_REQUESTED** [review]
I don't get how you were able to add this class and check that everything works fine, when there's code that references classes that don't exist on your branch.

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java (hunk 8 above)
Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java:55]: I simply add all the necessary files from the outside and then only commit the class that has to go to pull request.
All these "outsiders" files will be stashed in my local repo.

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java:55]: Of course the files I use are:

- https://github.com/Cogno-IDU/killer-app
- https://github.com/Cogno-IDU/smslibrary

Comment (giovannivelludo) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java:55]: Why don't you just pull from Cogno-IDU/killer-app/master to AlbertoUrsino/killer-app/AppManager before making any changes, so that you get a functioning app in your branch?
I don't exactly understand why you used git stash to put aside the "outsider" files needed to have a working app: from what I understand stashing is used when you're working on a branch and you have to switch to another, but you don't want to lose the changes you made to your working directory and you also don't want to commit them.

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java:55]: If I merge all the "outsider" files in my master and then call a pull request.. Well, I tried but in the pull request will being show all the "ousiders" files as "changed files".. I don't know how to resolve this problem, I tried with "cherry-commits" but I didn't understand them so I opted for this method and it works.

Comment (giovannivelludo) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java:55]: > Well, I tried but in the pull request will being show all the "ousiders" files as "changed files"

Could you elaborate on this? From what branch and into what branch did you pull (let's use the term "origin" to refer to your GitHub fork of killer-app, and "upstream" to refer to Cogno-IDU's killer-app)?
Maybe upload a video on YouTube showing what happens in Android Studio when you merge upstream/master into your local branch AppManager, then push to origin/AppManager, then make a pull request from AlbertoUrsino:AppManager to Cogno-IDU:master?

Comment (Cogno-Marco) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java:55]: What should happen is that you update the submodule from the commit it's on to a new commit based on the central repo. Since you're updating these you'll see the smslibrary submodule files as changed (since they in fact are changed). This is not a problem if they get in the pull request, you just have to update us that you updated the submodule, so we know that any file related to smslibrary is already ok since we reviewed it in it's own repository

**Comment by albertoursino** [comment]
In the last commit there is also the RingtoneHandler class that I moved to the right "structure" package.

**Review by giovannivelludo — APPROVED** [review]

**Comment by Cogno-Marco** [comment]
Ready to merge?

**Review by tonin-alessandra — APPROVED** [review]

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            ringtoneHandler.playRingtone(ringtone);
+            //Timer: the ringtone is playing for TIMEOUT_TIME seconds.
+            new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    ringtoneHandler.stopRingtone(ringtone);
+                }
+            }, TIMEOUT_TIME);
+        } else {
+            Toast.makeText(context, WRONG_PASSWORD_TOAST, Toast.LENGTH_SHORT).show();
```
Comment (albertoursino) [inline app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: This toast does not allow us to test the class on "onRingCommandReceived" method and in addition it is not nice to have a toast inside a class and not in the mainactivity.

What do u think?

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Untitled

What about this solution? Now we can do tests and also there's a better correctness of the code...

**Review by albertoursino — COMMENTED** [review]
I've done this modifications in the commit  aae0877

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java (hunk 9 above)
Comment (Cogno-Marco) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Beautiful solution changing from a toast to an exception. I would make just a little edit, from `WrongPasswordException` to `InvalidPasswordException`. Same meaning, different way to say it, which in my opinion is more inline with the standard java api (think about `IllegalArgumentException`)

Comment (tonin-alessandra) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: I think this is a better solution than toast, but I also agree with @Cogno-IDU  about the exception name.

Comment (CremaLuca) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: I don't think `IllegalPasswordException` is the name we're looking for, this code checks if the password is the same, not if it's valid / with the right format. Anyway the name is still kinda intuitive and both can be used, but I'd prefer `Wrong`

ps. I don't like at all the use of the adversative "but" in @tonin-alessandra comment: you are linking two positive sentences with a "but"! That's not correct! Should have used an "and"

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: After receiving another revision, perhaps from @giovannivelludo, I will fix these expressions.

Comment (Cogno-Marco) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: @CremaLuca I also chose `Illegal` as a keyword because of the specifications
> Exception thrown when the password received is not valid

Not valid, as in, Invalid

In the end it comes down to personal preferences but I'll just make it consistent

**Comment by albertoursino** [comment]
I did this 9fc23df commit because I tried on another PC to checkout this branch and control if everything was fine, but there were some errors caused by "smslibrary" which was not updated.

Now it should be all 100% working, I also merged the central repo with this commit c367ca0 in order to resolve some conflicts.

Try this branch in your computer and control if it works fine.

File: app/src/androidTest/java/com/eis/phoneringer/structure/AppManagerTest.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Before
+    public void init() {
+        context = InstrumentationRegistry.getInstrumentation().getTargetContext();
+        appManager = AppManager.getInstance();
+        passwordManager = new PasswordManager(context);
+    }
+
+    @Test
+    public void onRingCommandReceived_wrong_password() {
```
Comment (CremaLuca) [inline app/src/androidTest/java/com/eis/phoneringer/structure/AppManagerTest.java:31]: What about using
`@Test(expected = WrongPasswordException.class)`
and remove the try-catch? The current method of doing assert.fail to assert that exceptions are thrown is good only if exceptions are NOT Runtime, so when you **must** catch them. In this case, where the exception is runtime, we can use the expected.

File: app/src/androidTest/java/com/eis/phoneringer/structure/AppManagerTest.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Test
+    public void onRingCommandReceived_right_password() {
+        passwordManager.setPassword(VALID_PASSWORD);
+        try {
+            appManager.getInstance().onRingCommandReceived(context, new RingCommand(new SMSPeer(VALID_PHONE_NUMBER), VALID_PASSWORD), ringtoneHandler.getDefaultRingtone(context));
+            //Success
+        } catch (WrongPasswordException e) {
+            Assert.fail("Should hasn't thrown the WrongPasswordException");
```
Comment (CremaLuca) [inline app/src/androidTest/java/com/eis/phoneringer/structure/AppManagerTest.java]: Test automatically fails if an exception is thrown, no need for this catch

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+    /**
+     * If the password of the RingCommand received is valid then play ringtone for fixed amount of time
+     *
+     * @param context     of the application
+     * @param ringCommand received
+     * @param ringtone    A valid ringtone to be played
+     * @throws WrongPasswordException Exception thrown when the password received is not valid
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
```
Comment (CremaLuca) [inline app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: 1. Why is this required? 
2. What happens if someone has KitKat? Not saying that you should handle that case, but probably if this method doesn't work and there is no alternative method for lower APIs, the whole app doesn't make sense either. It'd probably be good to say that minimum API required is 5.0 on `build.gradle` and remove this annotation

**Review by CremaLuca — CHANGES_REQUESTED** [review]
Nice, mostly good except for a little clarification on handling exceptions in test methods and a weird RequiresApi that I don't understand which line requires it.

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java (hunk 9 above)
Comment (CremaLuca) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Agreed on `Ìllegal`

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java (hunk 12 above)
Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Method `playRingtone` of the class `RingtoneHandler` requires that API

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: I'm fixing that problem with a modification in the file `build.gradle` as you said

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: API 5.0 doesn't exist. Minimun level is 7...

Comment (CremaLuca) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Man, 5.0/Lollipop is build no. 21 https://source.android.com/setup/start/build-numbers

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: lvl 21 maybe

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Sorry I was thinking about the levels not the Api version

**Review by CremaLuca — APPROVED** [review]
Now it's perfect

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java (hunk 9 above)
Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Well, now you are in 3 that you want `Illigal` instead of `Wrong` so i'll fix it now.

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+            //Timer: the ringtone is playing for TIMEOUT_TIME seconds.
+            new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    ringtoneHandler.stopRingtone(ringtone);
+                }
+            }, TIMEOUT_TIME);
+        } else {
+            throw new WrongPasswordException();
+        }
```
Comment (Cogno-Marco) [inline app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: A quick suggestion (even though it's not necessary, you might still find this useful)
There's a way of programming where you focus on having the parameters you want.
Basically code like this
```
if(condition){
    //code
}
else {
    return;
}
```
Becomes
```
if( !condition){
    return;
}
//code
```

This way long nested if chains just become linear
```
if( c1){
    if( c2){
        if( c3){
            //code
        }
    }
}
```
becomes
```
if( !c1)
   return;
if( !c2)
    return;
if( !c3)
    return;

//code
```

In your case it would be
```
if ( !checkPassword(context, ringCommand)) throw new WrongPasswordException();

//code inside the if
```

Just a good way to write code in my opinion.

**Review by Cogno-Marco — APPROVED** [review]
Nothing to say, good work 👍

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java (hunk 13 above)
Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Do you mean something like that?

Untitled

Comment (albertoursino) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: I can change if you prefer.
In my opinion you are in right @Cogno-IDU, this _code writing method_ significantly speeds up both the method and, in the long run, the application itself.


Also makes the code even nicer...

**Comment by albertoursino** [comment]
Well, now I think you can merge it

File: app/src/main/java/com/eis/phoneringer/structure/AppManager.java (hunk 13 above)
Comment (Cogno-Marco) [reply app/src/main/java/com/eis/phoneringer/structure/AppManager.java]: Nah you don't need to change it. I just thought you didn't know about it, or maybe you didn't pay attention, so I just gave some feedback 😄 .
Also yeah, it's a pretty useful coding style, it even has it's own name. Something like... defensive programming?
